### ISSUE TITLE
"Double echo" in line editor mode under Emacs

### ISSUE THREAD

**Issue by maconcg** [gambit/gambit#921]
Hi,

I've noticed a line editor mode quirk that I believe was introduced with commit 556a884d.  When I evaluate an expression at the REPL, both the expression and the value are printed where only the value should be printed.  Example:

```
> (+ 1 1)
(+ 1 1)
2
```

This happens with `emacs -q` and in several modes (including term mode, shell mode, CMU scheme mode, and geiser).

I've been able to semi-fix the quirk by modifying [these lines of os_tty.c](https://github.com/gambit/gambit/blob/898028d1a5cdf9ffe9a9dc6d9b2257d00e18fe3d/lib/os_tty.c#L5320-L5329).  My semi-fix is to just prevent the execution of `d->lineeditor_mode = LINEEDITOR_MODE_SCHEME;` under Emacs (e.g., by wrapping the #ifdef in `else { ... }`).  It's only a _semi_-fix because this breaks any REPL run from Emacs' term mode.

To implement a "real" fix, I considered adding some granularity to `lineeditor_under_emacs` by checking for the presence of `comint` in the value of `$INSIDE_EMACS` (rather than just checking whether that environment variable is set). That way, lineediting_mode could be disabled only in comint-derived modes.  This would probably require the definition of a new `___env_var_contains_UCS_2` function(?).

My more immediate problem, however, is that I don't understand the intent behind `lineeditor_under_emacs` in its current form.  Under what Emacs mode would one get the correct result by running a REPL where lineeditor_mode = 1 and input_echo = 0?

Also: there's a good chance I've misunderstood some things.  I may not even be asking the right questions.  If that's the case, just prod me in the right direction.

**Comment by lassik** [comment]
#488 may be related

**Comment by maconcg** [comment]
> #488 may be related

Could be.  Typescript outputs (created via `script -c <gsi>`) are attached, plus visual-format versions for people who don't have `vis(1)` installed.  These were run with a gsi built from git with no local modifications.

`term-mode` output is identical to the output from an actual xterm terminal window, and each line ends with a carriage return. In `shell-mode`, the carriage returns are not output.

[repl-typescripts.tgz](https://github.com/user-attachments/files/16848041/repl-typescripts.tgz)

**Comment by feeley** [comment]
I had not noticed the double echo when running `emacs -q`.

Gambit has its own implementation of a line editor that is convenient when running from a terminal. It supports several emacs keybindings, matching parenthesis "flashing", etc. To do these things the tty must be put in nonblocking and nonecho mode. This causes issues when running under emacs because emacs is expecting the interactive program to echo the lines entered by the user. I believe emacs has a heuristic that if the first line received from the subprocess after entering some input line is identical to that input, it will silently eat that line (which prevents the double echo). Perhaps the escape codes for making the echo bold are interfering with that heuristic. Moreover this heuristic can cause problems with Scheme when entering a value which is printed by the REPL as the same text, for example entering "42".

I'm not an emacs expert, so it isn't clear what is the best behaviour for an interactive program running under emacs to avoid double echo and silently dropped output. If we can figure that out, then that behaviour can be implemented in os_tty.c .

**Comment by maconcg** [comment]
I have a few ideas I want to try.  One piece that might help is the `___DEBUG_TTY` functionality.  If I just stick a `#define ___DEBUG_TTY` at the top of the file, a subsequent `make` fails because `___printf` is not defined.

What is the intended way to enable/use those debug print statements?

**Comment by feeley** [comment]
You need to `./configure --enable-debug-log` for `___printf` to be defined. You can also add `--enable-debug` for general debugging features.

**Comment by lassik** [comment]
In Emacs `M-x shell` I get double-printing behavior. The shell command `tty` shows that Emacs does allocate a pseudo-teminal, but `echo $TERM` is `dumb`.

@feeley Are you testing for `dumb`? That would be an easy fix.

**Comment by lassik** [comment]
`M-x shell-command` with the command line `tty; echo $TERM` outputs:

```
not a tty
dumb
```

**Comment by maconcg** [comment]
You can change the TERM value set by emacs via `(setq comint-terminfo-terminal <type>)`.  For `dumb`, the double-printing is present and escape codes _are not_ output. For some other value, like `dumb-emacs-ansi` or `ansi`, the double-printing is present and escape codes _are_ output.  You can notice this at a glance because the gambit version line is highlighted when TERM has a non-`dumb` value.  I confirmed this `script`/`vis`.

From this I conclude that bold/color escape sequences can be ruled out.

I haven't managed to identify any behavior for any configuration that changes depending on whether `input_echo` is 0 or 1.  This value does appear to differ according to whether INSIDE_EMACS is defined, though (I see the appropriate `d->input_echo = ` lines).

**Comment by Negdayen** [comment]
I vaguely remember encountering this issue, my workaround was to create a custom mode for when I ran (run-scheme "gsc") in Emacs.  Not sure if this provides any new insight.

```
(define-derived-mode inferior-gambit-scheme-mode
  inferior-scheme-mode
  "inferior-gambit-scheme"
  "Major mode for interacting with an inferior Gambit Scheme process."
  ;; Something in recent combinations of Gambit (4.9.5) / Emacs (29.1) causes comint to
  ;; echo the input twice, this removes the duplicated input.
  (setq-local comint-process-echoes t)
  (setq-local comint-prompt-regexp "^[0-9]*> "))
```

**Comment by maconcg** [comment]
> I vaguely remember encountering this issue, my workaround was to create a custom mode for when I ran (run-scheme "gsc") in Emacs. Not sure if this provides any new insight.
> 
> ```
> (define-derived-mode inferior-gambit-scheme-mode
>   inferior-scheme-mode
>   "inferior-gambit-scheme"
>   "Major mode for interacting with an inferior Gambit Scheme process."
>   ;; Something in recent combinations of Gambit (4.9.5) / Emacs (29.1) causes comint to
>   ;; echo the input twice, this removes the duplicated input.
>   (setq-local comint-process-echoes t)
>   (setq-local comint-prompt-regexp "^[0-9]*> "))
> ```

Yep, in fact I was using a similar workaround for a bit.  `comint-process-echoes` is a handy workaround but can cause other problems (like having your input "eaten" when you evaluate something that evaluates to itself).

The comment says 4.9.5, but were you by chance running a newer version from git?  My experience was that this workaround wasn't needed with a Gambit built from the release tarball.

I'm hoping the patch I'm referencing will make it so there's no need to change the Emacs defaults.  With the patch applied, this work as (I think) they should on the setups I've tested (OpenBSD and macOS, graphical as well as Terminal-based Emacs), but there are lots configurations out there I'm not able to test on my own.  So if you have time to try the patch and it works, let me know, and if it doesn't, I'll try to fix it.

**Comment by Negdayen** [comment]
Yeah, I probably didn't include the git commit.  I will give the patch try and let you know the behavior I observe, I agree that Emacs/comint is your line editor in this scenario.

I'm not sure why I haven't been able to produce the eaten input behavior on Linux 6.6 / Emacs 29.4 / Gambit v4.9.5-83-g7d462916.

**Comment by maconcg** [comment]
To anyone reading who hasn't also checked #922: After some help from Negdayen, I think the patch for this is in a good place.  If you use Emacs and have the inclination, please give those patches a test and let me know if they fix or break your setup.  Testing could be helpful even if you don't use Emacs, in which case behavior should be identical with/without the patches.

**Comment by maconcg** [comment]
Ok, that makes sense -- can't just use `INSIDE_EMACS` as a blanket override.  Thanks.

**Comment by monnier** [comment]
Side note: This discussion is a bit unclear for me because people often talk about whether `gsi` runs "inside Emacs", but this is probably not the right way to think about it.  E.g. running `gsi` inside Emacs will likely behave differently in `shell-mode` or `scheme-interaction-mode` than in `term-mode` or in `libvterm`.

For that reason, as a general rule, I don't think `INSIDE_EMACS` is a useful env-var either to decide how to behave.  It's usually preferable to make such decisions based on `TERM`.

**Comment by maconcg** [comment]
For the benefit of anyone coming to this form elsewhere (and to avoid splitting the conversation among two threads): my reply is in #922.
